File: ClientMobile/app/src/main/java/com/healthapp/clientmobile/services/SensorService.java
package com.healthapp.clientmobile.services;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.healthapp.clientmobile.R;
import com.healthapp.clientmobile.sensors.SensorData;

import java.util.ArrayList;
import java.util.Date;


public class SensorService extends Service implements SensorEventListener {
    private static final String TAG = "AccelerometerService";
    private final ArrayList<SensorData> accelerometerDataList = new ArrayList<>();
    private SensorManager sensorManager;
    private Sensor sensor;
    private Handler handler;
    private Runnable runnable;

    private String userId;

    @Override
    public void onCreate() {
        super.onCreate();
//        calAidApp = (CalAidApp) getApplicationContext();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "onStartCommand: Service started");
        accelerometerDataList.clear();
        if(intent != null){
            if(intent.getAction() == "startAccService"){
                        handler = new Handler();
                        runnable = new Runnable() {
                            @Override
                            public void run() {
                                Log.d("CALAIDAPP -Acc service", "acc");
                                sendDataToDB();
                                handler.postDelayed(this, 10000);
                            }
                        };
                        handler.postDelayed(runnable, 10000);
                    }
                sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);

                sensor = sensorManager.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION);
                sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_NORMAL);


                final String CHANNELID = "Foreground Service ID";
                createNotificationChannel(CHANNELID);
                Notification.Builder notification = new Notification.Builder(this, CHANNELID)
                        .setContentText("Collecting accelerometer data...")
                        .setContentTitle("")
                        .setSmallIcon(R.drawable.icon_launcher);

                startForeground(1001, notification.build());
            }
            if(intent.getAction() == "stopAccService"){
                stopForeground(true);
                stopSelf();

        }

        return START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, "onDestroy: Service destroyed");
        super.onDestroy();
        sensorManager.unregisterListener(this);
        handler.removeCallbacks(runnable);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onSensorChanged(SensorEvent sensorEvent) {
        Log.d(TAG,+ sensorEvent.values[0] + " " + sensorEvent.values[1] + " " + sensorEvent.values[2]);
        SensorData data = new SensorData();
        data.setUserId(userId);
        data.setX(sensorEvent.values[0]);
        data.setY(sensorEvent.values[1]);
        data.setZ(sensorEvent.values[2]);
        data.setTimestamp(new Date(System.currentTimeMillis()));
        accelerometerDataList.add(data);
    }

    private void sendDataToDB() {
        if (accelerometerDataList.size() > 0) {
            final ArrayList<SensorData> dataToSend = new ArrayList<>(accelerometerDataList);
            accelerometerDataList.clear();
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int i) {

    }

    private void createNotificationChannel(String CHANNEL_ID) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "CalAidApp";
            String description = "Collecting accelerometer data";
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel channel = new NotificationChannel(CHANNEL_ID, name, importance);
            channel.setDescription(description);
            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }

}

File: MobileApp/app/src/main/java/com/mc/mobileapp/SensorService.kt
package com.mc.mobileapp

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import android.os.IBinder
import android.util.Log
import androidx.core.app.NotificationCompat

import com.mc.mobileapp.domains.SensorData
import com.mc.mobileapp.retrofit.ContextInfo
import com.mc.mobileapp.retrofit.EncryptedSensorBatchPayload

import com.mc.mobileapp.retrofit.ISensorApiService
import com.mc.mobileapp.retrofit.RetrofitClient
import com.mc.mobileapp.retrofit.SensorDataEncrypted
import com.mc.mobileapp.utilities.AesEncryption
import kotlinx.coroutines.*

const val CHANNEL_ID = "SensorServiceChannel"
const val NOTIFICATION_ID = 1

class SensorService : Service(), SensorEventListener {

    private lateinit var sensorManager: SensorManager
    private var accelerometer: Sensor? = null
    private var gyroscope: Sensor? = null
    private val coroutineScope = CoroutineScope(Dispatchers.IO)
    private val sensorApiService = RetrofitClient.create(ISensorApiService::class.java)
    private val sensorDataBuffer = mutableListOf<SensorData>()
    private var userId: Int = -1

    override fun onCreate() {
        super.onCreate()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Sensor Service",
                NotificationManager.IMPORTANCE_LOW
            )
            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }

        val notification: Notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("Sensor Service")
            .setContentText("Logging accelerometer and gyroscope data.")
            .setSmallIcon(android.R.drawable.ic_menu_compass)
            .build()

        startForeground(NOTIFICATION_ID, notification)

        sensorManager = getSystemService(SENSOR_SERVICE) as SensorManager
        accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        gyroscope = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)

        accelerometer?.let {
            sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL)
        }

        gyroscope?.let {
            sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL)
        }

        coroutineScope.launch {
            userId = getUserId()
            while (true) {
                delay(5000)
                sendSensorDataBatch()
            }
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        sensorManager.unregisterListener(this)
        coroutineScope.cancel()
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onSensorChanged(event: SensorEvent?) {
        event?.let {
            val timestamp = System.currentTimeMillis()
            val sensorData = when (event.sensor.type) {
                Sensor.TYPE_ACCELEROMETER -> SensorData(
                    x = event.values[0],
                    y = event.values[1],
                    z = event.values[2],
                    timestamp = timestamp,
                    sensorType = "accelerometer",
                    userId = userId
                )

                Sensor.TYPE_GYROSCOPE -> SensorData(
                    x = event.values[0],
                    y = event.values[1],
                    z = event.values[2],
                    timestamp = timestamp,
                    sensorType = "gyroscope",
                    userId = userId
                )

                else -> null
            }
            sensorData?.let { data ->
                synchronized(sensorDataBuffer) {
                    sensorDataBuffer.add(data)
                }
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

    private suspend fun sendSensorDataBatch() {
        val dataToSend: List<SensorData>
        synchronized(sensorDataBuffer) {
            dataToSend = ArrayList(sensorDataBuffer)
            sensorDataBuffer.clear()
        }

        if (dataToSend.isNotEmpty()) {
            try {
                AesEncryption.loadKeyFromAssets(this.applicationContext)

                val encryptedDataList = dataToSend.map { entry ->
                    val (xEnc, ivX) = AesEncryption.encrypt(entry.x.toString())
                    val (yEnc, ivY) = AesEncryption.encrypt(entry.y.toString())
                    val (zEnc, ivZ) = AesEncryption.encrypt(entry.z.toString())

                    SensorDataEncrypted(
                        sensorType = entry.sensorType,
                        timestamp = entry.timestamp,
                        userId = entry.userId,
                        x = xEnc,
                        y = yEnc,
                        z = zEnc,
                        ivX = ivX,
                        ivY = ivY,
                        ivZ = ivZ
                    )
                }

                val payload = EncryptedSensorBatchPayload(
                    data = encryptedDataList,
                    timestamp = System.currentTimeMillis(),
                    context = ContextInfo(
                        user_id = userId.toString(),
                        encryption = "aes"
                    )
                )

                sensorApiService.uploadSensorData(payload)

            } catch (e: Exception) {
                Log.e("SensorService", "Encryption or upload failed: ${e.localizedMessage}")
                e.printStackTrace()
            }
        }
    }

    private suspend fun getUserId(): Int {
        val sharedPreferences = getSharedPreferences("SHARED_PREFS", MODE_PRIVATE)
        val email = sharedPreferences.getString("email", null)

        return email?.let {
            MainActivity.database.userDao().getUserId(it) ?: -1
        } ?: -1
    }

}